lib/Solver/STPBuilder.cpp

```diff
@@ -465,26 +465,29 @@ ExprHandle STPBuilder::getInitialRead(const Array *root, unsigned index) {
   return vc_readExpr(vc, getInitialArray(root), bvConst32(32, index));
 }
 
-::VCExpr STPBuilder::getArrayForUpdate(const Array *root, 
+::VCExpr STPBuilder::getArrayForUpdate(const Array *root,
                                        const UpdateNode *un) {
+  ::VCExpr un_expr;
+  std::vector<const UpdateNode *> update_nodes;
+  for (; un && !_arr_hash.lookupUpdateNodeExpr(un, un_expr);
+       un = un->next.get()) {
+    update_nodes.emplace_back(un);
+  }
   if (!un) {
-      return getInitialArray(root);
+    un_expr = getInitialArray(root);
   }
-  else {
-      // FIXME: This really needs to be non-recursive.
-      ::VCExpr un_expr;
-      bool hashed = _arr_hash.lookupUpdateNodeExpr(un, un_expr);
-      
-      if (!hashed) {
-        un_expr =
-            vc_writeExpr(vc, getArrayForUpdate(root, un->next.get()),
-                         construct(un->index, 0), construct(un->value, 0));
 
-        _arr_hash.hashUpdateNodeExpr(un, un_expr);
-      }
-      
-      return un_expr;
+  for (std::size_t i = update_nodes.size(); i > 0;) {
```

Comment:
Maybe use an iterator approach here with `update_nodes.rbegin()`. I would love to have a reversed range-based for-loop in such cases.

Reply:
That is available [in C++20](https://en.cppreference.com/w/cpp/ranges/reverse_view) or slightly less pretty [using LLVM](https://llvm.org/doxygen/classllvm_1_1iterator__range.html). Since we are still two C++ versions behind, I used the LLVM variant here.